phoenix-core/src/it/java/org/apache/phoenix/end2end/MutationBatchFailedStateMetricWithDeleteAndUpsertIT.java

```diff
@@ -0,0 +1,181 @@
+        Map<String, String> clientProps = Maps.newHashMapWithExpectedSize(2);
+        clientProps.put(QueryServices.COLLECT_REQUEST_LEVEL_METRICS, String.valueOf(true));
+        clientProps.put(QueryServices.TRANSACTIONS_ENABLED, "true");
+        setUpTestDriver(new ReadOnlyProps(serverProps.entrySet().iterator()),
+                new ReadOnlyProps(clientProps.entrySet().iterator()));
+
+    }
+
+    @Parameterized.Parameters(name = "MutationBatchFailedStateMetricWithDeleteAndUpsertIT_transactionProvider={0}")
+    public static synchronized Collection<Object[]> data() {
+        return Arrays.asList(new Object[][] { { "OMID" }, { null } });
+    }
+
+
+    @Test
+    public void testFailedUpsertAndDelete() throws Exception {
+        Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+        Connection conn = null;
+
+        try {
+            conn = DriverManager.getConnection(getUrl(), props);
+
+            Statement stmt = conn.createStatement();
+            // adding coprocessor which basically overrides prebatchmutate whiich is called for all
+            // the mutations
+            // it is called before applying mutation to region
+            TestUtil.addCoprocessor(conn, deleteTableName,
+                    ImmutableIndexIT.DeleteFailingRegionObserver.class);
+            // trying to delete 4 rows with this single delete statement
+            String dml = String.format("DELETE FROM %s where val1 >  1", deleteTableName);
+            stmt.execute(dml);
+
+            String upsertSQL = String.format(upsertStatement, deleteTableName);
+            PreparedStatement preparedStatement = conn.prepareStatement(upsertSQL);
+            preparedStatement.setString(1, "ROW_6");
+            preparedStatement.setInt(2, 6);
+            preparedStatement.setInt(3, 12);
+            preparedStatement.execute();
+
+            conn.commit();
+
+            throw new AssertionError("Commit should not have succeeded");
+        } catch (SQLException e) {
+            Map<String, Map<MetricType, Long>> mutationMetrics =
+                    conn.unwrap(PhoenixConnection.class).getMutationMetrics();
+            int mfs =
+                    mutationMetrics.get(deleteTableName).get(MUTATION_BATCH_FAILED_SIZE).intValue();
+            int dbfs =
+                    mutationMetrics.get(deleteTableName).get(DELETE_BATCH_FAILED_SIZE).intValue();
+            int upfs =
+                    mutationMetrics.get(deleteTableName).get(UPSERT_BATCH_FAILED_SIZE).intValue();
+            long gfs =
+                    GLOBAL_MUTATION_BATCH_FAILED_COUNT.getMetric().getValue();
+
+
+            Assert.assertEquals(4, dbfs);
+            Assert.assertEquals(5, mfs);
+            Assert.assertEquals(1, upfs);
+            //for the second parameter the global mutation batch failed count is double the original parameter because
+            //global metrics do not reset to 0 for each parameter
+            if(this.transactionProvider == null){
```

Comment:
Likewise we're assuming an order here when we really care about whether this is the first or second parameter.